File: Garbage/ShadoMath/ShadoMath/Fraction.cpp
#include <string>
#include <stdio.h>
#include <vector>
#include <iostream>
#include "Fraction.h"

/*************************
**************************
*** Shado Fraction Lib.***
**************************
************************/

Fraction::Fraction(int a, int b) {
	numerator = a;
	denumerator = b;
}

Fraction::Fraction(float decimal) {

	int tries = 0;
	float num = decimal;

	while (num - (int)num != 0) {
		num *= 10;
		tries++;
	}

	numerator = (int)num;

	for (int i = 0; i < tries; i++) {
		denumerator *= 10;
	}
}

int Fraction::getNumerator() {
	return numerator;
}

int Fraction::getDumerator() {
	return denumerator;
}

Fraction Fraction::inverse() {
	return Fraction(denumerator, numerator);
}

void Fraction::simplify() {

	for (int i = 10; i > 0; i--) {
		if (numerator % i == 0 && denumerator % i == 0) {
			numerator = numerator / i;
			denumerator = denumerator / i;
		}
	}


}

std::string Fraction::toString() {
	return std::to_string(numerator) + " / " + std::to_string(denumerator);
}

void Fraction::print() {
	std::cout << this->toString();
}

Fraction Fraction::add(Fraction& fracB) {

	// Find common denumerator
	Fraction commonA = Fraction(numerator * fracB.denumerator, denumerator * fracB.denumerator);
	Fraction commonB = Fraction(fracB.numerator * denumerator, fracB.denumerator * denumerator);

	Fraction result = Fraction(commonA.numerator + commonB.numerator, commonA.denumerator);
	result.simplify();

	return result;
}

Fraction Fraction::substract(Fraction& fracB) {
	// Find common denumerator
	Fraction commonA = Fraction(numerator * fracB.denumerator, denumerator * fracB.denumerator);
	Fraction commonB = Fraction(fracB.numerator * denumerator, fracB.denumerator * denumerator);

	Fraction result = Fraction(commonA.numerator - commonB.numerator, commonA.denumerator);
	result.simplify();

	return result;
}

Fraction Fraction::multiply(float number) {
	return Fraction((numerator * number) / denumerator);
}

Fraction Fraction::multiply(Fraction& fracB) {
	return Fraction(numerator * fracB.numerator, denumerator * fracB.denumerator);
}

Fraction Fraction::devide(Fraction& fracB) {
	Fraction copy = fracB.inverse();

	return (*this).multiply(copy);
}

Fraction Fraction::devide(float number) {
	return Fraction((numerator / number) / denumerator);
}

bool Fraction::equals(Fraction& fracB) {
	return (numerator / denumerator) == (fracB.numerator / fracB.denumerator);
}

// Overload operators for Fraction
Fraction Fraction::operator+(Fraction fracB) {
	return (*this).add(fracB);
}

Fraction Fraction::operator-(Fraction fracB) {
	return (*this).substract(fracB);
}

Fraction Fraction::operator*(Fraction fracB) {
	return (*this).multiply(fracB);
}

Fraction Fraction::operator/(Fraction fracB) {
	return (*this).devide(fracB);
}


std::ostream& operator<<(std::ostream& os, Fraction frac)
{
	os << frac.toString();
	return os;
}

File: COMP-249/Lectures/src/ShadoMath/Fraction.java
/***
 *
 * Shado Fraction class library
 *
 * */

package ShadoMath;

public class Fraction {

	private int numerator;
	private int denominator;

	/***
	 *
	 * @param numerator The numerator of the fraction
	 * @param denominator The denominator of the fraction
	 */
	public Fraction(int numerator, int denominator) {
		this.numerator = numerator;
		this.denominator = denominator;
	}

	/***
	 *
	 * @param number the decimal number you want to convert to a fraction
	 */
	public Fraction(double number) {

		double tempNum = number;
		int tries = 0;

		while (tempNum - (int) tempNum != 0.0) {
			tempNum *= 10;
			tries++;
		}

		this.numerator = (int) tempNum;
		this.denominator = (int) Math.pow(10, tries);

		this.simplify();

	}

	/***
	 * Copy constructor
	 * @param other The fraction you want to copy
	 */
	public Fraction(final Fraction other) {
		this(other.numerator, other.denominator);
	}

	/***
	 * Simplifies the calling Fraction
	 */
	public void simplify() {
		for (int i = 10; i > 0; i--) {
			if (this.numerator % i == 0 && this.denominator % i == 0) {
				this.numerator = this.numerator / i;
				this.denominator = this.denominator / i;
			}
		}
	}

	/***
	 * Inverses the calling Fraction
	 */
	public void inverse() {
		int temp = this.numerator;
		this.numerator = this.denominator;
		this.denominator = temp;
		this.simplify();
	}

	/***
	 * Multiplies the fraction by -1
	 */
	public void opposite() {
		this.numerator = -1 * this.numerator;
	}

	/***
	 * Adds a Fraction to the calling Fraction
	 * @param other The fraction you want to add to the calling Fraction
	 * @return Returns the result of the addition
	 */
	public Fraction add(final Fraction other) {
		Fraction commonA = new Fraction(this.numerator * other.denominator, this.denominator * other.denominator);
		Fraction commonB = new Fraction(other.numerator * this.denominator, other.denominator * this.denominator);

		Fraction result = new Fraction(commonA.getNumerator() + commonB.getNumerator(), commonA.getDenominator());
		result.simplify();
		return result;
	}

	/***
	 * Subtracts a Fraction to the calling Fraction
	 * @param other The fraction you want to subtract to the calling Fraction
	 * @return Returns the result of the subtraction
	 */
	public Fraction subtracts(final Fraction other) {
		other.opposite();
		return this.add(other);
	}

	/***
	 * Multiplies a Fraction to the calling Fraction
	 * @param other The fraction you want to multiply to the calling Fraction
	 * @return Returns the result of the multiplication
	 */
	public Fraction multiply(final Fraction other) {
		return new Fraction(this.numerator * other.getNumerator(), this.denominator * other.getDenominator());
	}

	/***
	 * Divides a Fraction to the calling Fraction
	 * @param other The fraction you want to divide to the calling Fraction
	 * @return Returns the result of the division
	 */
	public Fraction divide(final Fraction other) {
		other.inverse();
		return this.multiply(other);
	}

	/***
	 * Raises the numerator and to denominator of the Fraction to a exponent
	 * @param exponent The specific exponent
	 * @return Returns the result of the power
	 */
	public Fraction power(int exponent) {
		return new Fraction((int) Math.pow(this.numerator, exponent), (int) Math.pow(this.denominator, exponent));
	}

	// Getters

	/***
	 * @return Returns the numerator of the power
	 */
	public int getNumerator() {
		return this.numerator;
	}

	/***
	 * @return Returns the denominator of the power
	 */
	public int getDenominator() {
		return this.denominator;
	}

	// String and print

	/***
	 * @return Returns "numerator / denominator"
	 */
	public String toString() {
		return String.format("%d / %d", this.numerator, this.denominator);
	}

	/***
	 * prints the calling Fraction to the console
	 */
	public void print() {
		System.out.println(this.toString());
	}
}
